### ISSUE TITLE
Extremely High Loss, which quickly goes to *0* by the End; Configs that Were Fine Are Broken Now

### ISSUE THREAD

**Issue by e-p-armstrong** [axolotl-ai-cloud/axolotl#1806]
### Please check that this issue hasn't been reported before.

- [X] I searched previous [Bug Reports](https://github.com/axolotl-ai-cloud/axolotl/labels/bug) didn't find any similar reports.

### Expected Behavior

The same configs should produce a model similar to what it had done before the update around July 30. There should be no dramatic instability with the loss values.

### Current behaviour

Whereas before the loss started around 2.3 and proceeded to be something around 0.8 by the end of a 5-epoch training run, now, with the same config, it starts at 4.2 (!) and goes to 0 in places by the end. Accordingly, the resulting model is overfit to hell and braindead.

In this WandB graph, the blue thing was the loss graph of a model before this problem was introduced. The ones that spike in the middle are all ones after the fact, with slightly different settings as I tried to figure out what was going on.

![image](https://github.com/user-attachments/assets/ecb34749-bff6-4096-8781-ce792e26acd6)

![image](https://github.com/user-attachments/assets/52e5f067-4a54-40b2-9a41-28de81bf1d39)

### Steps to reproduce

Train a LoRA with the given config and some data. Observe obscenely high loss values and overfitting where there was none before.

### Config yaml

```yaml
base_model: abacusai/Llama-3-Smaug-8B
model_type: LlamaForCausalLM
tokenizer_type: AutoTokenizer

load_in_8bit: false
load_in_4bit: false
strict: false

datasets:
  - path: json
    data_files: ./training_data_2/final_annotation_convs.jsonl
    type: chat_template
    field_messages: conversations
    message_field_role: from
    message_field_content: value
    roles:
      user:
        - human
      assistant:
        - gpt
      system:
        - system
  - path: json
    data_files: ./training_data_2/rephrase.jsonl
    type: chat_template
    field_messages: conversations
    message_field_role: from
    message_field_content: value
    roles:
      user:
        - human
      assistant:
        - gpt
      system:
        - system
  
dataset_prepared_path: ./training_datafolder
output_dir: ./training_out

adapter: lora

sequence_len: 4096
sample_packing: true
pad_to_sequence_len: true
shuffle_merged_datasets: true

lora_r: 128
lora_alpha: 256
lora_dropout: 0.20
lora_target_linear: true
lora_fan_in_fan_out:
lora_target_modules:
  - gate_proj
  - down_proj
  - up_proj
  - q_proj
  - v_proj
  - k_proj
  - o_proj

wandb_project: fixed-cleaned-dropout-and-neftune
wandb_entity:
wandb_watch:
wandb_run_id:
wandb_log_model:

gradient_accumulation_steps: 4
micro_batch_size: 2
eval_batch_size: 1
num_epochs: 5
optimizer: paged_adamw_8bit
lr_scheduler: cosine
learning_rate: 0.0002
weight_decay: 0
# Gradient clipping max norm
max_grad_norm: 1.0
noisy_embedding_alpha: 5

train_on_inputs: false
group_by_length: false
bf16: true
fp16: false
tf32: false

gradient_checkpointing: unsloth
early_stopping_patience:
resume_from_checkpoint: 
local_rank:
logging_steps: 1
xformers_attention:
flash_attention: true

# fsdp:
  # - full_shard
  # - auto_wrap
# fsdp_config:
  # fsdp_offload_params: false
  # fsdp_state_dict_type: FULL_STATE_DICT
  # fsdp_transformer_layer_cls_to_wrap: LlamaDecoderLayer
# warmup_steps: 10
warmup_steps: 0
auto_resume_from_checkpoints: false
#warmup_ratio: 0.5
eval_steps: 10
saves_per_epoch: 1
eval_sample_packing: false
save_total_limit: 2
debug:
deepspeed: deepspeed_configs/zero2.json
special_tokens:
  pad_token: "<|end_of_text|>"
```

### Possible solution

Potentially useful for finding a reason: rolling back axolotl versions and doing a new pip install -r requirements does NOT seem to fix it, at least with the version I am rolling back to. So maybe this is an issue with an underlying dependency? I have no idea how to address it though, or how to find which dep is the problem.

IDK if axolotl is the problem here, but as it stands, I can't train anything anymore. Would appreciate support in knowing what settings to turn off to fix this, until those deps fix their darn problems. Very concerned about this bug becoming a permenant thing in case it only happens during edge cases.

### Which Operating Systems are you using?

- [X] Linux
- [ ] macOS
- [ ] Windows

### Python Version

3.11.9

### axolotl branch-commit

main/78b42a3fe13c49e317bc116b9999c30e070322cc

### Acknowledgements

- [X] My issue title is concise, descriptive, and in title casing.
- [X] I have searched the existing issues to make sure this bug has not been reported yet.
- [X] I am using the latest version of axolotl.
- [X] I have provided enough information for the maintainers to reproduce and diagnose the issue.

**Comment by e-p-armstrong** [comment]
NOTE just confirmed this appears to be happening on runpod specifically. I get drastically different tokens/device, as well as loss, on vast.ai vs runpod. Specifically runpod seems to have twice the number of tokens as vast. And 20 steps vs 5.

Any ideas about what might be going on here? Completely lost, I'm using the same docker image on both providers so why is one "works on my machine" while the other is a dumpster fire?

Vast:
![image](https://github.com/user-attachments/assets/6dd9d4be-5266-4065-bc23-998461111a38)

RP:
![image](https://github.com/user-attachments/assets/4d15ede0-92a8-4383-a5b2-520d3c10ce40)

**Comment by e-p-armstrong** [comment]
UPDATE the issue MAY be with the transformers library; reverting ALL THE WAY BACK to e4063d60a793d97630ee12dcd728c05fbb496d88 has so far resulted in a more stable loss curve. Will update if the model comes out sane; I've had a stable loss curve that made the model braindead before.

**Comment by e-p-armstrong** [comment]
Updating: reverting to e406d3 PARTLY ALLEVIATES, **BUT DOES NOT SOLVE**, the issue. Requesting help.

**Comment by sklyar61** [comment]
I also noticed issues with training: two weeks ago, I successfully trained the model. The pipeline was fine-tuned. Yesterday, I trained the model again with all the same inputs as before, and the model stopped correctly stopping the generation.

I provided more details here: #1812

**Comment by e-p-armstrong** [comment]
@sklyar61 How did you manage to make the model stop, using the new version of axolotl? Was it removing special_tokens: pad token: <|eot_id|>?

**Comment by sklyar61** [comment]
It seems that in the new version (at least in the latest docker-container winglian/axolotl:main-latest), I was unable to stop the model generation.

The last successful training I conducted was on July 22, 2024.

At that time, I used
specialtokens:
  padtoken: 
      <|end_of_text|>
And it worked.

The docker images from August 5 and 6, 2024, no longer worked.

I looked at the axolotl code, and it seems there is an addition of special tokens, which made my confusion even greater.

Maybe someone can tell me which code I need to check so that I can make changes?

**Comment by sklyar61** [comment]
Is anyone else experiencing issues with stopping the generation in the latest versions of Axolotl?

**Comment by qeternity** [comment]
Yep same issues here, have not been able to even partially resolve this.

**Comment by e-p-armstrong** [comment]
@qeternity @sklyar61 
Managed to solve. Needs a rollback.

One issue might be that if you turn flash attention off, Axolotl will just break. The loss will start out absurdly high. This setting is broken; FA must be on. I spent about 2 days isolating this problem.

Another, is that a recent axolotl version changed the default chat template, if you're manually specifying it. So in the config I shared, which didn't specify it for each dataset (note the absence of a `chat_template` field), once axolotl updated the template was set to something else that the model wasn't trained on, and the clash of chat templates caused problems. If you were relying on consistent defaults this might have caught you out.

Finally, quality issues can happen with hte most recent axolotl version or maybe it's caused by one of its dependencies. I have recreated this with example data using the settings shown in this issue originally -- this is an issue with the code, not the data. Furthermore, the number of training steps can be inconsistent and very small in recent axolotl versions. I do not know why this is, but something is broken.

Either way, you can solve this:

1. Use the axolotl docker image
2. cd into axolotl/
3. git fetch --unshallow
4. checkout to e6b299dd79f75537f3e247a3d22dfed5d3885bfb
5. pip install -e '.[flash-attn, deepspeed]'
6. Get a config from before the issue, like this one:
```
Babel==2.15.0
GitPython==3.1.43
Jinja2==3.1.4
Markdown==3.6
MarkupSafe==2.1.5
PyYAML==6.0.1
Pygments==2.18.0
Send2Trash==1.8.3
Werkzeug==3.0.3
absl-py==2.1.0
accelerate==0.32.0
addict==2.4.0
aiofiles==23.2.1
aiohttp==3.9.5
aioitertools==0.11.0
aiosignal==1.3.1
altair==5.3.0
annotated-types==0.7.0
anyio==4.4.0
argon2-cffi-bindings==21.2.0
argon2-cffi==23.1.0
arrow==1.3.0
art==6.2
asttokens==2.4.1
async-lru==2.0.4
attrs==23.2.0
autoawq==0.2.6
autoawq_kernels==0.0.7
beautifulsoup4==4.12.3
bitsandbytes==0.43.1
bleach==6.1.0
cachetools==5.4.0
causal-conv1d==1.4.0
certifi==2024.7.4
cffi==1.16.0
charset-normalizer==3.3.2
click==8.1.7
cmake==3.30.1
colorama==0.4.6
coloredlogs==15.0.1
comm==0.2.2
contourpy==1.2.1
cycler==0.12.1
datasets==2.19.1
debugpy==1.8.2
decorator==5.1.1
deepspeed==0.14.3
defusedxml==0.7.1
dill==0.3.8
dnspython==2.6.1
docker-pycreds==0.4.0
docstring_parser==0.16
docutils==0.16
einops==0.8.0
email_validator==2.2.0
evaluate==0.4.1
executing==2.0.1
fastapi-cli==0.0.4
fastapi==0.111.1
fastcore==1.6.1
fastjsonschema==2.20.0
ffmpy==0.3.2
filelock==3.15.4
fire==0.6.0
flash-attn==2.6.2
fonttools==4.53.1
fqdn==1.5.1
frozenlist==1.4.1
fschat==0.2.36
fsspec==2024.3.1
galore-torch==1.0
gcsfs==2024.3.1
gitdb==4.0.11
google-api-core==2.19.1
google-auth-oauthlib==1.2.1
google-auth==2.32.0
google-cloud-core==2.4.1
google-cloud-storage==2.18.0
google-crc32c==1.5.0
google-resumable-media==2.7.1
googleapis-common-protos==1.63.2
gradio==3.50.2
gradio_client==0.6.1
grpcio==1.65.1
h11==0.14.0
hf_transfer==0.1.8
hjson==3.1.0
httpcore==1.0.5
httptools==0.6.1
httpx==0.27.0
huggingface-hub==0.24.2
humanfriendly==10.0
idna==3.7
importlib_resources==6.4.0
iniconfig==2.0.0
ipykernel==6.29.5
ipython==8.26.0
ipywidgets==8.1.3
isoduration==20.11.0
jedi==0.19.1
jmespath==1.0.1
joblib==1.4.2
json5==0.9.25
jsonpointer==3.0.0
jsonschema-specifications==2023.12.1
jsonschema==4.23.0
jupyter-events==0.10.0
jupyter-lsp==2.2.5
jupyter_client==8.6.2
jupyter_core==5.7.2
jupyter_server==2.14.2
jupyter_server_terminals==0.5.3
jupyterlab==4.2.4
jupyterlab_pygments==0.3.0
jupyterlab_server==2.27.3
jupyterlab_widgets==3.0.11
kiwisolver==1.4.5
latex2mathml==3.77.0
lion-pytorch==0.1.2
llvmlite==0.43.0
lomo-optim==0.1.1
mamba-ssm==1.2.0.post1
markdown-it-py==3.0.0
markdown2==2.5.0
matplotlib-inline==0.1.7
matplotlib==3.9.1
mdurl==0.1.2
mistune==3.0.2
mpmath==1.3.0
multidict==6.0.5
multiprocess==0.70.16
nbclient==0.10.0
nbconvert==7.16.4
nbformat==5.10.4
nest-asyncio==1.6.0
networkx==3.3
nh3==0.2.18
ninja==1.11.1.1
notebook==7.2.1
notebook_shim==0.2.4
numba==0.60.0
numpy==1.26.4
nvidia-cublas-cu12==12.1.3.1
nvidia-cuda-cupti-cu12==12.1.105
nvidia-cuda-nvrtc-cu12==12.1.105
nvidia-cuda-runtime-cu12==12.1.105
nvidia-cudnn-cu12==8.9.2.26
nvidia-cufft-cu12==11.0.2.54
nvidia-curand-cu12==10.3.2.106
nvidia-cusolver-cu12==11.4.5.107
nvidia-cusparse-cu12==12.1.0.106
nvidia-ml-py==12.555.43
nvidia-nccl-cu12==2.20.5
nvidia-nvjitlink-cu12==12.5.82
nvidia-nvtx-cu12==12.1.105
oauthlib==3.2.2
optimum==1.16.2
orjson==3.10.6
overrides==7.7.0
packaging==23.2
pandas==2.2.2
pandocfilters==1.5.1
parso==0.8.4
peft==0.11.1
pexpect==4.9.0
pillow==10.4.0
pip==24.1.2
platformdirs==4.2.2
pluggy==1.5.0
prometheus_client==0.20.0
prompt_toolkit==3.0.47
proto-plus==1.24.0
protobuf==4.25.4
psutil==6.0.0
ptyprocess==0.7.0
pure_eval==0.2.3
py-cpuinfo==9.0.0
pyarrow-hotfix==0.6
pyarrow==17.0.0
pyasn1==0.6.0
pyasn1_modules==0.4.0
pycparser==2.22
pydantic==2.6.3
pydantic_core==2.16.3
pydub==0.25.1
pynvml==11.5.3
pyparsing==3.1.2
pytest==8.3.2
python-dateutil==2.9.0.post0
python-dotenv==1.0.1
python-json-logger==2.0.7
python-multipart==0.0.9
pytz==2024.1
pyzmq==26.0.3
referencing==0.35.1
regex==2024.7.24
requests-oauthlib==2.0.0
requests==2.32.3
responses==0.18.0
rfc3339-validator==0.1.4
rfc3986-validator==0.1.1
rich==13.7.1
rpds-py==0.19.1
rsa==4.7.2
s3fs==2024.3.1
s3transfer==0.10.2
safetensors==0.4.3
scikit-learn==1.2.2
scipy==1.14.0
semantic-version==2.10.0
sentencepiece==0.2.0
sentry-sdk==2.11.0
setproctitle==1.3.3
setuptools==69.5.1
shellingham==1.5.4
shortuuid==1.0.13
shtab==1.7.1
six==1.16.0
smmap==5.0.1
sniffio==1.3.1
soupsieve==2.5
stack-data==0.6.3
starlette==0.37.2
svgwrite==1.4.3
sympy==1.13.0
tensorboard-data-server==0.7.2
tensorboard==2.17.0
termcolor==2.4.0
terminado==0.18.1
threadpoolctl==3.5.0
tiktoken==0.7.0
tinycss2==1.3.0
tokenizers==0.19.1
toolz==0.12.1
torch-optimi==0.2.1
tornado==6.4.1
tqdm==4.66.4
traitlets==5.14.3
transformers==4.43.1
triton==2.3.1
trl==0.9.6
typer==0.12.3
types-python-dateutil==2.9.0.20240316
typing_extensions==4.12.2
tyro==0.8.5
tzdata==2024.1
uri-template==1.3.0
urllib3==2.2.2
uvicorn==0.30.3
uvloop==0.19.0
wandb==0.17.5
watchfiles==0.22.0
wavedrom==2.0.3.post3
wcwidth==0.2.13
webcolors==24.6.0
webencodings==0.5.1
websocket-client==1.8.0
websockets==11.0.3
wheel==0.43.0
widgetsnbextension==4.0.11
wrapt==1.16.0
xformers==0.0.27
xxhash==3.4.1
yarl==1.9.4
zstandard==0.22.0
```
Get it onto the docker thing, and then do pip install -r requirements.txt

You should now be able to train without random unexplained quality loss in the final model. This quality loss is a separate issue from the loss spike itself, I have found them to be independent.

Further, I have no idea if it is axolotl or an upstream dependency causing this problem with quality, but the fact that simply rolling back was not enough, and that using a different requirements.txt is needed, seems to imply it is not solely axolotl's fault. Maybe transformers? Pytorch? Deepspeed?

Either way, the newest version of *something* is broken. @winglian thoughts?

**Comment by winglian** [comment]
@e-p-armstrong what happens if you disable sample packing?

**Comment by e-p-armstrong** [comment]
Hey! Good question re: sample packing, will test tomorrow and get back to you (night where I'm at).

**Comment by qeternity** [comment]
No, sample packing does not make a difference.

**Comment by qeternity** [comment]
Above does not help either.

**Comment by qeternity** [comment]
Ok so for me it turns out I was missing the root level `chat_template` value (although I did have it additionally at the dataset level).

```
chat_template: llama3
datasets:
```

**Comment by e-p-armstrong** [comment]
@winglian The loss spike might have been solved, either by background dependency updates or chat_template or sample packing being turned off, I do not know; but the quality issues that form the most significant part of this issue sadly remain even when sample packing is turned off. I do not run into these same issues when using old versions of axolotl with frozen dependencies.

**Comment by e-p-armstrong** [comment]
Here is loss vs steps in case it is interesting.
![image](https://github.com/user-attachments/assets/8e2ba23b-e746-4f88-b81f-5397001e830d)

**Comment by e-p-armstrong** [comment]
Loss vs steps in case it is interesting: 
![image](https://github.com/user-attachments/assets/a8b2c126-7fad-4207-99a2-b28c91fa2089)

**Comment by qeternity** [comment]
Ok so if I manually build the training set and use `type: input_output` for the dataset it seems to be ok. The loss still looks too low, but I have kicked off a small training job.

@winglian there seems to be something wrong with the chat templating. I'll dig a bit deeper when I have time.
